base/compiler/typeinfer.jl

```diff
@@ -1007,48 +1020,131 @@ function typeinf_frame(interp::AbstractInterpreter, mi::MethodInstance, run_opti
     return frame
 end
 
+# N.B.: These need to be aligned with the C side headers
+"""
+    SOURCE_MODE_NOT_REQUIRED
+
+Indicates to inference that the source is not required and the only fields
+of the resulting `CodeInstance` that the caller is interested in are types
+and effects. Inference is still free to create a CodeInstance with source,
+but is not required to do so.
+"""
+const SOURCE_MODE_NOT_REQUIRED = 0x0
+
+"""
+    SOURCE_MODE_ABI
+
+Indicates to inference that it should return a CodeInstance that can
+either be `->invoke`'d (because it has already been compiled or because
+it has constabi) or one that can be made so by compiling its `->inferred`
+field.
+
+N.B.: The `->inferred` field is volatile and the compiler may delete it.
+In such a case, it will first set the `invoke` field to a method that
+will block the thread until compilation is completed.
+"""
+const SOURCE_MODE_ABI = 0x1
+
+"""
+    SOURCE_MODE_FORCE_SOURCE
+
+Indicates that inference must always produce source in the `->inferred` field.
+This may mean that inference will need to re-do inference (if the `->inferred`
+field was previously deleted by the JIT) or may need to synthesize source for
+other kinds of CodeInstances.
+
+N.B.: The same caching considerations as SOURCE_MODE_ABI apply.
+"""
+const SOURCE_MODE_FORCE_SOURCE = 0x2
+
+"""
+    SOURCE_MODE_FORCE_SOURCE_UNCACHED
+
+Like `SOURCE_MODE_FORCE_SOURCE`, but ensures that the resulting code instance is
+not part of the cache hierarchy, so the `->inferred` field may be safely used
+without the possibility of deletion by the compiler.
+"""
+const SOURCE_MODE_FORCE_SOURCE_UNCACHED = 0x3
+
+function ci_has_source(code::CodeInstance)
+    inf = @atomic :monotonic code.inferred
+    return isa(inf, CodeInfo) || isa(inf, String)
+end
+
+"""
+    ci_has_abi(code::CodeInstance)
+
+Determine whether this CodeInstance is something that could be invoked if we gave it
+to the runtime system (either because it already has an ->invoke ptr, or because it
+has source that could be compiled).
+"""
+function ci_has_abi(code::CodeInstance)
+    ci_has_source(code) && return true
+    return code.invoke !== C_NULL
```

Comment:
These are monotonic loads, so you've introduced a data-race here also. I think you wanted an `:acquire` on one of these (may this one) to fix that, so that it must observe this being set. And also switch the order of the checks, so that it correctly pairs with the release store.